### ISSUE TITLE
~28% of all the network evals are duplicates?

### ISSUE THREAD

**Issue by ywrt** [gcp/leela-zero#459]
tl;dr: I think leela training games can easily be sped up by ~40%

Hacking in a duplicate tracker into Network::get_scored_moves() indicates that many calls to network to score the state have recently been evaluated.

I added something like: 
```
Network::get_scored_moves(...)
     [...]
     gather_features(state, planes);
 
+    std::vector<bool> bits;
+    for (const auto& p : planes)
+      for (unsigned int i = 0; i < p.size(); ++i)
+        bits.push_back(p[i]);
+    auto hash = std::hash<std::vector<bool>>{}(bits);
+    {
+      LOCK(tried_mutex, lock);
+      if (tried.count(hash)) ++tried_dup;
+      ++tried_total;
+      tried.insert(hash);
+    }
```

and after running a game with "-p 1600", it gave tried_dup=158745, tried_total=556768 for a duplicate rate of about 28.5%

As best as I can tell, most of these duplicates come from the node tree being expanded from scratch on every move.

For example, when UCTSearch::think() is called, the first thing it does is expand out the child nodes for the new root node. But that node was almost certainly already expanded for the previous move, so this becomes a duplicate call to Network::get_scored_moves().  etc, etc for a decent portion of the previous tree.

I suspect the right thing to do here is to enable re-using UCTSearch objects, and when a move it played it finds the appropriate child node, unlinks it (and it's children) from the existing tree, and the makes it the new root node.

**Comment by bood** [comment]
Did #315 address this already?

**Comment by ywrt** [comment]
I didn't see that. Which is silly, because I just reimplemented the same feature. 
https://github.com/gcp/leela-zero/compare/master...ywrt:reuse-nodes?expand=1
Eh. 

It turns out to be better, but not perfect:  It drops the duplicate calls from ~ 28% down to ~7%

**Comment by gcp** [comment]
Consider rebasing to next, I can't merge any of that code as is. Given that there is still a discrepancy, maybe the Ttable should just store the network evaluations even if it has to be made a bit smaller?

**Comment by ywrt** [comment]
I re-implemented this by just caching the network eval results. Based against next.
